src/lib/Behat/PageElement/ElementFactory.php

```diff
@@ -0,0 +1,42 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace EzSystems\EzPlatformAdminUi\Behat\PageElement;
+
+use EzSystems\EzPlatformAdminUi\Behat\Helper\UtilityContext;
+
+class ElementFactory
+{
+    /**
+     * Creates a Element object based on given Element Name.
+     *
+     * @param UtilityContext $context
+     * @param string $elementName Name of the Element to creator
+     *
+     * @return AdminList|Dialog|RightMenu|UpperMenu|UpdateForm|Breadcrumb|Notification Element to interact with
+     */
+    public static function createElement(UtilityContext $context, string $elementName, ?string ...$parameters): Element
+    {
+        switch ($elementName) {
+            case AdminList::ELEMENT_NAME:
+                return new AdminList($context, $parameters[0]);
+            case Dialog::ELEMENT_NAME:
+                return new Dialog($context);
+            case RightMenu::ELEMENT_NAME:
+                return new RightMenu($context);
+            case UpperMenu::ELEMENT_NAME:
+                return new UpperMenu($context);
+            case UpdateForm::ELEMENT_NAME:
+                return new UpdateForm($context);
+            case Breadcrumb::ELEMENT_NAME:
+                return new Breadcrumb($context);
+            case Notification::ELEMENT_NAME:
```

Comment:
This is not very SOLID, I'm also not a huge fan of static factory methods. If any new element appears, this list needs to be updated and maintained. The problem here is that we're a little bit limited due to lack of DI container.

if I were to change it - since the switch cases are really simple - I'd just avoid using factory in favor of creating of objects in-place. It actually would produce less code :-)

POV @Nattfarinn @webhdx

Reply:
Actually i had the same opinion, but when i start working on automated tests, there was already PageObjectFactory, and after discussion we decided to stay with this concept. I'll be happy to discuss it again, but if it's possible, as a new task.